rusoto/core/src/request.rs

```diff
@@ -96,42 +95,25 @@ impl fmt::Debug for BufferedHttpResponse {
     }
 }
 
-/// Future returned from `HttpResponse::buffer`.
-pub struct BufferedHttpResponseFuture {
-    status: StatusCode,
-    headers: HeaderMap<String>,
-    future: ::futures::stream::Concat2<ByteStream>,
-}
-
-impl Future for BufferedHttpResponseFuture {
-    type Item = BufferedHttpResponse;
-    type Error = HttpDispatchError;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        self.future
-            .poll()
-            .map_err(std::convert::Into::into)
-            .map(|r#async| {
-                r#async.map(|body| BufferedHttpResponse {
-                    status: self.status,
-                    headers: mem::replace(&mut self.headers, Default::default()),
-                    body,
-                })
-            })
-    }
-}
-
 impl HttpResponse {
     /// Buffer the full response body in memory, resulting in a `BufferedHttpResponse`.
-    pub fn buffer(self) -> BufferedHttpResponseFuture {
-        BufferedHttpResponseFuture {
-            status: self.status,
-            headers: self.headers,
-            future: self.body.concat2(),
+    pub async fn buffer(self) -> Result<BufferedHttpResponse, HttpDispatchError> {
```

Comment:
I'm not sure a `BufferedHttpResponse` _needs_ to exist, as you could instead return a `Result<http::Response<bytes::Bytes>, HttpDispatchError>`.

(Broadly, I think I'd rather expose the underlying `http` crate as much as possible in `rusoto/core`, but I'm just this dude with some [not necessarily correct!] opinions.)